### PR TITLE
fix: improve JWKS caching

### PR DESCRIPTION
**Pull request by sdelamo**
Close: https://github.com/micronaut-projects/micronaut-security/issues/771

### CODE REVIEW

**Review by Steven-M-Brown — COMMENTED** [review]
I don't think this actually fixes the issues in #771 as it still considers each set of jwk's independently due to the `@EachBean(JwksSignatureConfiguration.class)`.

I think this might actually be more broken now in as it won't ever refresh on a miss seemingly.

Consider this:
Key doesn't match anything in the jwk set: will return false out of verify.
Key does match but fails verification (not sure when this would actually even happen): triggers refresh by nuking the set and recurses.

The key not matching anything is the one that needs to trigger the refresh to handle a singing key rotation (signed by a new key id, causes a no-match) but it needs to consider all available JWKSets across all available `JwksSignature` before doing the refresh (original issue when more than one source for jwk's is used).

On an unrelated note I think your double-checked locking usage in `getJWKSet` is incorrect (missing volatile on `jwkSet`).

**Comment by Steven-M-Brown** [comment]
If it would be helpful I can try and clean up our bandaid fix and put up a PR just so you can get an idea of what amounted to a bit of a hack to solve it.

**Comment by sdelamo** [comment]
> B) Key does match but fails verification (not sure when this would actually even happen): triggers refresh by nuking the set and recurses.

I think that could happen when an authorization server rotates key without changing the key id. That should not be done by an authorization server but it is an scenario.

**Comment by Steven-M-Brown** [comment]
> I think that could happen when an authorization server rotates key without changing the key id. That should not be done by an authorization server but it is an scenario.
Okay that makes sense I guess (despite feeling really wrong to do I'm sure it happens).

**Comment by sdelamo** [comment]
I have added a test and code to handle scenario C. 

https://github.com/micronaut-projects/micronaut-security/pull/823/commits/07e55a694425a3224b1c691c6a40b8cf86ec2eab

Let me know what you think @Steven-M-Brown

**Comment by sdelamo** [comment]
I have added a test for no `kid`. https://github.com/micronaut-projects/micronaut-security/pull/823/commits/96fb057a4e2cdd677a8543f768a88168c9658d00 If I think that covers A. Not sure what else we can do about it. What do you think @Steven-M-Brown ?

**Comment by sdelamo** [comment]
someone could attack the server by sending JWT signed multiple times. Such attack may cause the server to hit the remote JWKS constantly. To avoid that I have added a configuration property set to 60 seconds. 

`'micronaut.security.token.jwt.signatures.jwks.*.cache-expiration`. If the token verification fails, The JWKS are not refreshed from remote server until the cache has expired.

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 1]
```diff
@@ -152,6 +184,19 @@ public boolean verify(SignedJWT jwt) throws JOSEException {
...
+        }
+    }
+
+    @Override
+    public boolean isJwksCachePresent() {
+        return this.jwkSet != null;
+    }
+
     /**
      * Instantiates a JWKSet for a give url.
```
Comment (Steven-M-Brown) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:201]:
```suggestion
     * Instantiates a JWKSet for a given url.
```

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 2]
```diff
@@ -182,30 +226,60 @@ protected JWKSet loadJwkSet(String url) {
...
      */
+    @Deprecated
     protected List<JWK> matches(SignedJWT jwt, @Nullable JWKSet jwkSet, int refreshKeysAttempts) {
+        return matches(jwt, jwkSet);
+    }
+
+    /**
+     * Calculates a list of JWK matches for a JWT.
+     *
+     * Since the JWTSet is cached it attempts to refresh it (by calling its self recursive)
```
Comment (Steven-M-Brown) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:240]:
```suggestion
     * Since the JWKSet is cached it attempts to refresh it (by calling itself recursively)
```

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 3]
```diff
@@ -47,17 +53,16 @@
...
+    @Deprecated
     public static final int DEFAULT_REFRESH_JWKS_ATTEMPTS = 1;
 
     private static final Logger LOG = LoggerFactory.getLogger(JwksSignature.class);
-
     private final JwkValidator jwkValidator;
-
+    private final JwksSignatureConfiguration jwksSignatureConfiguration;
+    private Date jwkSetCachedAt;
     private JWKSet jwkSet;
```
Comment (Steven-M-Brown) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:65]: I still think this needs a `volatile` to be correct.

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java (hunk 2 above)
Comment (Steven-M-Brown) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:240]: Also it doesn't seem to call itself recursively anymore, is this just an out of date comment?

**Review by Steven-M-Brown — COMMENTED** [review]
I'm finding the diff a tad hard to follow so I may have to pull down the branch and read it that way. There are a few minor grammar/javadoc mistakes and a couple possible race conditions plus a missing volatile but it looks okay as far as I can tell at this point.

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java (hunk 2 above)
Comment (sdelamo) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:238]: I have remove the line from the java doc. Please, note that I have deprecated the method.

Comment (sdelamo) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java]: thanks, I removed it.

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 4]
```diff
@@ -104,6 +129,13 @@ public JwksSignature(String url,
         return getJWKSet().map(JWKSet::getKeys).orElse(Collections.emptyList());
     }
 
+    @Override
+    @NonNull
+    public Optional<Boolean> isJwksCacheExpired() {
+        return (jwkSetCachedAt == null) ? Optional.empty() :
+                Optional.of(!new Date().before(jwkSetCachedAt));
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java]: I've added  `sychronized` to the method and also extracted a variable.

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java (hunk 3 above)
Comment (sdelamo) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java]: I've added `volatile` to both.

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 5]
```diff
@@ -131,9 +131,27 @@ public Integer getCacheExpiration() {
...
+    @Override
+    public synchronized boolean isJwksCachePresent() {
+        return this.jwkSet != null;
+    }
+
+    @Override
+    @NonNull
+    public synchronized Optional<List<String>> getJwksKeyIDs() {
+        JWKSet jwkSet = this.jwkSet;
+        return parseJwksKeyIDs(jwkSet);
```
Comment (Steven-M-Brown) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:154]: So while this does fix it I don't think it was the right fix as it adds *a ton* of potential contention between requests. Note that a few of these are used by `comparator` from `validate` which is in the hot path of requests.

Rather than a sledgehammer (synchronized) I would recommend a scalpel (store it in a local variable so it doesn't check the null then NPE a couple instructions later). In this case it is actually probably okay if it compares against slightly out of date data based on the intended usages in order to not block other requests.

**Comment by sdelamo** [comment]
@jameskleeh can you review this?

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 6]
```diff
@@ -104,6 +129,31 @@ public JwksSignature(String url,
         return getJWKSet().map(JWKSet::getKeys).orElse(Collections.emptyList());
     }
 
+    @Override
+    @NonNull
+    public synchronized Optional<Boolean> isJwksCacheExpired() {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:134]: I don't think this needs to be synchronized

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 7]
```diff
@@ -104,6 +129,31 @@ public JwksSignature(String url,
...
+        Date cachedAt = this.jwkSetCachedAt;
+        return isExpired(cachedAt);
+    }
+
+    @Override
+    public synchronized void clearJwksCache() {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:140]: I don't think this needs to be synchronized

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 8]
```diff
@@ -104,6 +129,31 @@ public JwksSignature(String url,
...
+        this.jwkSet = null;
+        this.jwkSetCachedAt = null;
+    }
+
+    @Override
+    public synchronized boolean isJwksCachePresent() {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:146]: I don't think this needs to be synchronized

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java (hunk 5 above)
Comment (jameskleeh) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:154]: Agreed I don't see the need for synchronization here

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 9]
```diff
@@ -169,43 +219,71 @@ protected JWKSet loadJwkSet(String url) {
...
+    private static Optional<List<String>> parseJwksKeyIDs(@Nullable JWKSet jwkSet) {
+        return jwkSet == null ? Optional.empty() : Optional.of(jwkSet.getKeys()
+                .stream()
+                .map(JWK::getKeyID)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList()));
+    }
+
+    @NonNull
+    private static Optional<Boolean> isExpired(@Nullable Date cachedAt) {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:283]: Should probably use Instant instead of Date

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/validator/JwtValidator.java [hunk 10]
```diff
@@ -191,38 +193,87 @@ private boolean hasAtLeastTwoDots(String token) {
...
+            Optional<JWT> optionalJWT = validate(jwt, config);
+            if (optionalJWT.isPresent()) {
+                return optionalJWT;
+            }
+        }
         return Optional.empty();
     }
 
-    private static Comparator<SignatureConfiguration> comparator(JWSAlgorithm algorithm) {
+    private static int compareMatchingKids(SignatureConfiguration sig,
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/validator/JwtValidator.java:251]: Is this supposed to mean `Key IDs`? If so I would have that has the name because what this method does isn't clear by the name

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/validator/JwtValidator.java [hunk 11]
```diff
@@ -191,38 +193,87 @@ private boolean hasAtLeastTwoDots(String token) {
...
-            } catch (final JOSEException e) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("Verification failed with signature configuration: {}, passing to the next one", config);
+        Optional<JWT> optionalJWT = validate(jwt, sortedConfigs);
+        if (optionalJWT.isPresent()) {
+            return optionalJWT;
+        }
+
+        // If any of the signature configurations is a JwksCache, evict the cache and attempt to verify again
+        for (SignatureConfiguration c : sortedConfigs.stream()
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/validator/JwtValidator.java:205]: Don't use a stream here. You can just check for instanceof in the loop

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java (hunk 5 above)
Comment (jameskleeh) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:154]: I think I take that back because we need synchronization because we have separate operations, check for null, get key set
